### PR TITLE
Widen to constants, fix termination check

### PR DESCRIPTION
**Pull request by elazarg**
1. Handle widening by once widening towards prededefined constant values.
2. Fix termination check by using a dedicated loop counter for each WTO cycle.

The handling of widening is somewhat ad-hoc. @caballa suggested another solution; the termination check is orthogonal, and may apply to any other solution as well, though it is simplistic too.

### CODE REVIEW

File: src/crab_verifier.cpp [hunk 1]
```diff
@@ -130,19 +122,16 @@ static void print_report(std::ostream& os, const checks_db& db, const Instructio
...
         }
     }
     os << "\n";
-    if (!db.maybe_nonterminating.empty()) {
-        os << "Could not prove termination on join into: ";
-        for (const label_t& label : db.maybe_nonterminating) {
-            os << label << ", ";
-        }
-        os << "\n";
+    constexpr int max_instructions = 100000;
```
Comment (dthaler) [inline src/crab_verifier.cpp:132]: FYI for comparison, I heard one of the linux maintainers say the max instructions for that verifier is 1 million

File: src/assertions.cpp [hunk 2]
```diff
@@ -218,6 +218,9 @@ class AssertExtractor {
             return { Assert{TypeConstraint{ins.dst, TypeGroup::number}} };
         }
     }
+    vector<Assert> operator()(IncrementLoopCounter ins) const {
+        assert(false);
+    }
```
Comment (dthaler) [inline src/assertions.cpp:223]: The windows build has an error saying something has to be returned, so maybe something like:
```suggestion
        return {};
    }
```

File: src/crab_verifier.cpp (hunk 1 above)
Comment (elazarg) [reply src/crab_verifier.cpp:132]: "instructions" is now a misnomer, as it only counts iterations. It will probably end up as a parameter.

**Comment by elazarg** [comment]
The important part is to decide whether this is the right way to do widening. 

The changes to [fwd_analyzer.cpp](
https://github.com/vbpf/ebpf-verifier/pull/479/files#diff-f17aa118aafd3fa8882c3c01b415a2f9e65aa2020d38700f9ac8791977c69c85) and its vicinity (ebpf_domain_t::limits()) need to be carefully reviewed. Maybe I should separate those changes to a specific commit.

Also, the change to the termination check means `instruction count` is a misnomer, should be "iteration count" or something.

**Comment by dthaler** [comment]
> Also, the change to the termination check means `instruction count` is a misnomer, should be "iteration count" or something.

What does "iteration count" mean?  E.g., for a nested loop like:
```
   for (i=0; i<2; i++) {
      for (j=0; j<2; j++) {
          ...
      }
   }
```
Is the iteration count == 4?

**Comment by elazarg** [comment]
There should be two counters there, each == 2. This example is not currently handled since I forgot to add the counters to inner cycles, and the current initialization is misplaced.

**Comment by dthaler** [comment]
Programs that don't widen to a constant still go to infinity and fail a termination check, such as:
```
SEC("xdp")
int
xdp_sum(xdp_md_t* ctx)
{
    uint64_t sum = 0;
    const uint8_t* p = (const uint8_t*)ctx->data;
    uint64_t len = ctx->data_end - ctx->data;

    for (int x = 0; x < len; x++)
        sum += p[x];

    bpf_printk("packet sum - %llu", sum);
    return XDP_PASS;
}
```

**Comment by absagar1** [comment]
This may be related; forward loops fail while reverse works:

```#define DATA_BUFFER_LEN (8 * 1024)

typedef struct _pkt_t
{
    uint16_t length;
    uint8_t data[DATA_BUFFER_LEN];
} pkt_t;

struct
{
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __type(key, uint32_t);
    __uint(value_size, sizeof(pkt_t));
    __uint(max_entries, 1);
} test_map SEC(".maps");

SEC("xdp/looptest")
int
xdp_loop_test(xdp_md_t* ctx)
{
    pkt_t* pkt;
    uint32_t key = 0;

    pkt = bpf_map_lookup_elem(&test_map, &key);
    if (pkt == NULL) {
        return XDP_DROP;
    }

    if (pkt->length > DATA_BUFFER_LEN) {
        return XDP_DROP;
    }

#if 1 // this case fails

    for (int i = 0; i < pkt->length; i += 1) {
        if (pkt->data[i] != 0) {
            return XDP_DROP;
        }
    }

#else // this case works

    for (int i = pkt->length - 1; i >= 0; i -= 1) { // while loop also works
        if (pkt->data[i] != 0) {
            return XDP_DROP;
        }
    }

#endif 

    return XDP_PASS;
}```

**Comment by caballa** [comment]
Regarding this program:

```
SEC("xdp")
int
xdp_sum(xdp_md_t* ctx)
{
    uint64_t sum = 0;
    const uint8_t* p = (const uint8_t*)ctx->data;
    uint64_t len = ctx->data_end - ctx->data;

    for (int x = 0; x < len; x++)
        sum += p[x];

    bpf_printk("packet sum - %llu", sum);
    return XDP_PASS;
}
```

Next, I show what I believe is the CFG generated by prevail. Note that I added some `assume` statements to add assumptions added by prevail. Ignore for now the comment`/* SOME MAGIC ASSUMPTION HERE */`:

```
block_start:
   havoc(packet_size)
   // These two assumptions are added by prevail
   assume(packet_size >= 0)
   assume(packet_size <= 65534):
  
   havoc(packet_start)
   // These two assumptions are added by prevail
   assume(packet_start >= 4098)
   assume(packet_start <= 2147418112)
  
   havoc(packet_end)
   // These two assumptions are added by prevail
   assume(packet_end <= packet_size)
   assume(packet_end >= 0)

   r1 := packet_end - packet_start
   /* SOME MAGIC ASSUMPTION HERE */
   goto block_header
block_header:
   r3 := 0
   r0 := 0
   goto body
block_body:
   r4 := packet_start
   r4 := r4 + r3
   // This is the check that prevail cannot prove 
   assert(r4 <= packet_size)
   r3 := r3 + 1
   if (r1>r3)  goto block_body else goto block_exit
block_exit:  ...
```

prevail cannot prove the assertion. The reason I believe is that we need to tell prevail that `packet_size` is not only between `0` and `65534` but also that `packet_size == packet_end - packet_start`. 
If we replace `/* SOME MAGIC ASSUMPTION HERE */` with `assume(packet_size == r1)` I do believe that prevail should prove the assertion in this program. 

@elazarg: let me know if you follow my explanation.  If yes, please could you add the extra assumption `packet_size == packet_end - packet_start` as part of the invariants added by prevail?  

Moreover, please we need to merge this PR. This is something that Dave really needs.
But let's see first if we can prove this program.

**Comment by caballa** [comment]
@elazar: you are right that the offset of the start of the packet is zero. 

Let me try again. This is the entailment test that Prevail cannot prove:

```
[
    meta_offset=[-4098, 0], 
    --> packet_size=[0, 65534]<--, 
    r0.type=number, 
    -->r1.svalue=[1, 65534]<--, 
    r1.svalue=r1.uvalue, r1.type=number, r1.uvalue=[1, 65534], 
    r10.stack_offset=512, r10.svalue=[512, 2147418112], r10.type=stack, 
    r2.packet_offset-packet_size<=0, r2.packet_offset=0, r2.svalue-r4.svalue<=0, 
    r2.svalue-r4.uvalue<=0, r2.svalue=[4098, 2147418112], r2.type=packet, 
    r3.svalue-r1.svalue<=-1, r3.svalue-r1.uvalue<=-1, r3.svalue-r4.svalue<=-4098, 
    r3.svalue-r4.uvalue<=-4098,   r3.svalue=[0, 65533], r3.svalue=r3.uvalue, r3.svalue=r4.packet_offset, 
    r3.type=number, r3.uvalue-r1.svalue<=-1, r3.uvalue-r1.uvalue<=-1, r3.uvalue-r4.svalue<=-4098, 
    r3.uvalue-r4.uvalue<=-4098, r3.uvalue=[0, 65533], r3.uvalue=r4.packet_offset, 
    -->r4.packet_offset-r1.svalue<=-1<---, 
    r4.packet_offset-r1.uvalue<=-1, r4.packet_offset-r4.svalue<=-4098, r4.packet_offset-r4.uvalue<=-4098,
    -->r4.packet_offset=[0, 65533]<--, 
    r4.svalue-r1.svalue<=2147418111, r4.svalue-r1.uvalue<=2147418111, r4.svalue-r2.svalue<=65533, 
    r4.svalue-r3.svalue<=2147418112, r4.svalue-r3.uvalue<=2147418112, r4.svalue-r4.packet_offset<=2147418112, 
    r4.svalue=[4098, 2147483645], r4.svalue=r4.uvalue, r4.type=packet, r4.uvalue-r1.svalue<=2147418111, 
    r4.uvalue-r1.uvalue<=2147418111, r4.uvalue-r2.svalue<=65533, r4.uvalue-r3.svalue<=2147418112, 
    r4.uvalue-r3.uvalue<=2147418112, r4.uvalue-r4.packet_offset<=2147418112, r4.uvalue=[4098, 2147483645]]
```

ENTAILS 

```
r4.packet_offset + -packet_size <= -1
```

I use `--> ... <--` to highlight the key invariants.

In the above invariants, `r1` is size of the packet. It's set at the entry of the program and it's not modified after.

```
entry:
  r2 = *(u32 *)(r1 + 0);
  ...
  r1 = *(u32 *)(r1 + 4);
  ...
  r1 -= r2;
```

 Crucially, Prevail infers that

`r4.packet_offset <= r1.svalue - 1` 

which is exactly what we need to prove the assertion.  However, the assertion uses `packet_size` but there is no relationship between `packet_size` and `r1`!

Note that `packet_size=[0, 65534]`. This includes the case where the
size of the packet is 0. But it cannot be 0 because the assertion check
happens inside the loop and we execute the loop only if the size of the packet is not 0.

I think that the problem is that Prevail adds as an initial invariant that packet_size=[0, 65534] but it needs also to add the initial invariant  

`packet_size = end_of_packet - start_of_packet`

where `end_of_packet = *(r1 + 0)` and `start_of_packet = *(r1+4)` at the
entry of the program when `r1` points to the packet.

**Comment by elazarg** [comment]
I think there are at least two issues:
1. PREVAIL does not set up the correct pre invariants, which may be related to @caballa's hypothesis; I had to add invariants to "forward loop" that weren't there for the compiled program.
2. Tests seem to depend on each other somehow; on my machine, running ` ./run_yaml test-data/loop.yaml` and running `./run_yaml test-data/loop.yaml "forward loop"` yield different invariants for "forward loop", specifically at the point on widening. It's not necessarily related, but I need to investigate it.

**Comment by elazarg** [comment]
I believe there's no issue with termination-checking specifically. There's an issue with modelling the pair of left-shift/right-shift that the compiler generates in certain cases. Whenever we can prove this pair is no-op (probably after implementing #63) we can keep relational information; otherwise, we lose it and cannot prove the relationship between registers and the size of the packet.

**Comment by dthaler** [comment]
@elazarg @caballa any progress on getting loops to work?

**Comment by caballa** [comment]
@dthaler: we are happy with this PR so it can be merged. 
To be clear, this PR focuses only on improving the widening operation (precision on loops)

There are two pending issues:

(1) Handle programs such as the loop program you shared with us on Jun 27th. We concluded that the problem with that example has nothing to do with loops. The problem is that there is a zero extension implemented as a pair of left shift and logical right that it's not currently modeled precisely by Prevail. This is related to this https://github.com/vbpf/ebpf-verifier/issues/63.

(2) Proving termination on loop programs. This requires more work.

Our plan is to finish the implementation of (1) by the end of next week. 

I know you have been waiting already long time for better support of loops but please give us until next week so that we can merge this PR and (1).  

Thank you for your patience.

**Comment by elazarg** [comment]
Merging. I'm on a reserve duty and won't have capacity in the near future.

**Comment by caballa** [comment]
Thanks @elazarg  and be safe!

File: src/crab/ebpf_domain.cpp [hunk 3]
```diff
@@ -909,8 +909,39 @@ ebpf_domain_t ebpf_domain_t::operator&(const ebpf_domain_t& other) const {
...
+        inv += r.svalue >= std::numeric_limits<int32_t>::min();
+        inv += r.uvalue <= std::numeric_limits<uint32_t>::max();
+        inv += r.uvalue >= 0;
+        inv += r.stack_offset <= EBPF_STACK_SIZE;
+        inv += r.stack_offset >= 0;
+        inv += r.shared_offset <= r.shared_region_size;
+        inv += r.shared_offset >= 0;
+        inv += r.packet_offset <= variable_t::packet_size();
+        inv += r.packet_offset >= 0;
+        if (thread_local_options.check_termination) {
```
Comment (dthaler) [inline src/crab/ebpf_domain.cpp:927]: This test will never be true when invoked by check.exe since this function is called before main() processes the command line.
